### PR TITLE
Copy the request ContentLength and TransferEncoding field in proxy

### LINKED ISSUE #16093: kube-apiserver HTTP proxy removes Content-Length header
Kubernetes' HTTP proxy removes the headers Content-Length and Transfer-Encoding from proxied requests. This breaks PUT and POST requests that require the Content-Length header and use chunked transfer encoding. I've created a producer for the problem with k8s configuration, Dockerfile and outputs at https://github.com/tiran/k8n-contentlength

In the reproducer example I'm making a curl POST request to a service via the proxy APO endpoint http://localhost:8080/api/v1/proxy/namespaces/default/services/contentlength/. Both curl -v and tcpdump show that curl is sending a Content-Length header. The simple socket server inside the pod gets all data except for the Content-Length header.

### PR DESCRIPTION
**Pull request by caesarxuchao**
Thanks to @tiran for identifying the problem and proposing the fix.

@lavalamp @liggitt 

fix #16093

### CODE REVIEW

File: pkg/apiserver/proxy.go [hunk 1]
```diff
@@ -151,6 +151,8 @@ func (r *ProxyHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	}
 	httpCode = http.StatusOK
 	newReq.Header = req.Header
+	newReq.ContentLength = req.ContentLength
+	newReq.TransferEncoding = req.TransferEncoding
```
Comment (liggitt) [inline pkg/apiserver/proxy.go]: whatever we do here we also need to do in generic/rest/proxy.go (a refactor to make apiserver/proxy.go delegate to the generic proxy is overdue, but should be a separate PR)

Comment (liggitt) [inline pkg/apiserver/proxy.go]: Looks like the reverse proxy strips the transfer encoding header.... isn't it a problem to send content length to the backend without transfer encoding?

Comment (liggitt) [inline pkg/apiserver/proxy.go]: in any case, I'd want tests through the proxy to a backend using no transport encoding, gzip, and chunked, with and without content-length, to make sure the backend gets a correct content length in all cases

Comment (liggitt) [inline pkg/apiserver/proxy.go]: hmm... this also gets set prior to the upgrade attempt... not sure if spdy or websockets would ever set a content length or transfer encoding, but we should at least check on that

Comment (caesarxuchao) [inline pkg/apiserver/proxy.go]: Thanks @liggitt. I'll take a look at the issues you mentioned tomorrow or Friday.

File: pkg/apiserver/proxy_test.go [hunk 2]
```diff
@@ -34,6 +36,115 @@ import (
...
+				contentLengthInHeader, err := strconv.Atoi(req.Header.Get("Content-Length"))
+				if err != nil {
+					t.Errorf("unexpected error: %v", err)
+				}
+				if e, a := expectedContentLength, int64(contentLengthInHeader); e != a {
+					t.Errorf("expected %v, got %v", e, a)
+				}
+			}
+
+			// Help wanted: why is the transfer-encoding header field always empty?
```
Comment (caesarxuchao) [inline pkg/apiserver/proxy_test.go]: @liggitt @lavalamp, the backend server always receives request with empty transfer-encoding header. Is this expected?

File: pkg/apiserver/proxy_test.go [hunk 3]
```diff
@@ -34,6 +36,115 @@ import (
...
+func containGzip(transferEncodings []string) bool {
+	for _, encoding := range transferEncodings {
+		if encoding == "gzip" {
+			return true
+		}
+	}
+	return false
+}
+
+func TestProxyRequestContentLengthAndTransferEncoding(t *testing.T) {
```
Comment (caesarxuchao) [inline pkg/apiserver/proxy_test.go]: @liggitt, please take a look at this test. I'm baffled by golang's http package :(

File: pkg/apiserver/proxy_test.go (hunk 2 above)
Comment (lavalamp) [inline pkg/apiserver/proxy_test.go]: Perhaps look at the header instead of the variable?

Comment (lavalamp) [inline pkg/apiserver/proxy_test.go]: Oh, you are... hm

Comment (caesarxuchao) [inline pkg/apiserver/proxy_test.go]: For the record, in the received request, the struct field req.TranferEncoding = []string{"chunked"} for the last 2 test cases. I think go library deduced this somehow.

File: pkg/apiserver/proxy_test.go [hunk 4]
```diff
@@ -34,6 +36,115 @@ import (
...
+		{[]string{"identity"}, "question", "default"},
+		{[]string{"chunked"}, "question", "default"},
+		// RFC2616 section-3.6: Whenever a transfer-coding is applied to a message-body, the set of
+		// transfer-codings MUST include "chunked", unless the message is terminated by closing the
+		// connection.
+		{[]string{"chunked", "gzip"}, "qqqqqqqqqquestion", "default"},
+	}
+
+	for _, item := range table {
+		proxyServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
```
Comment (lavalamp) [inline pkg/apiserver/proxy_test.go]: Suggest naming this "downstreamServer"-- it's not the proxy, it's the thing the proxy dials, right?

File: pkg/apiserver/proxy_test.go [hunk 5]
```diff
@@ -34,6 +36,115 @@ import (
...
+func calculateContentLength(reqBody string, transferEncodings []string) int64 {
+	if len(transferEncodings) == 0 || (len(transferEncodings) == 1 && transferEncodings[0] == "identity") {
+		return int64(len(reqBody))
+	}
+	// RFC2616 section 4.4: The Content-Length header field MUST NOT be sent if these two lengths
+	// are different (i.e., if a Transfer-Encoding header field is present).
+	return int64(-1)
+}
+
+func containGzip(transferEncodings []string) bool {
```
Comment (lavalamp) [inline pkg/apiserver/proxy_test.go]: shouldGzip?

File: pkg/apiserver/proxy_test.go [hunk 6]
```diff
@@ -34,6 +36,115 @@ import (
...
+		defer server.Close()
+
+		proxyTestPattern := "/api/version2/proxy/namespaces/default/foo/id/some/dir"
+		var reader io.Reader
+		if containGzip(item.transferEncodings) {
+			var b bytes.Buffer
+			gzw := gzip.NewWriter(&b)
+			if _, err := gzw.Write([]byte(item.reqBody)); err != nil {
+				t.Errorf("unexpected error: %v", err)
+			}
```
Comment (lavalamp) [inline pkg/apiserver/proxy_test.go]: It might work better if you Flush and/or Close gzw?

File: pkg/apiserver/proxy_test.go [hunk 7]
```diff
@@ -34,6 +36,115 @@ import (
...
+			"POST",
+			server.URL+proxyTestPattern,
+			reader,
+		)
+		if err != nil {
+			t.Errorf("unexpected error %v", err)
+			continue
+		}
+		req.TransferEncoding = item.transferEncodings
+		req.Header.Set("Transfer-Encoding", strings.Join(item.transferEncodings, ","))
```
Comment (lavalamp) [inline pkg/apiserver/proxy_test.go]: I think you should call Header.Add for each one instead of joining them manually. https://golang.org/pkg/net/http/#Header.Add

**Comment by lavalamp** [comment]
I think I found a couple bugs in the test, but I'm not sure if they'll actually fix your problem or not. Let me know.

File: pkg/apiserver/proxy_test.go (hunk 2 above)
Comment (caesarxuchao) [inline pkg/apiserver/proxy_test.go]: I prints the content in the buffer in the [conn](https://golang.org/src/net/http/server.go#L131) struct and [persistConn](https://golang.org/src/net/http/transport.go#L818) struct to understand what's written/read by the go standard library to/from the wire. The golang library only supports `chuncked` as a transfer-encoding, more specifically,

- When a request is sent either by a default golang http client or by a golang server, the only possible value in the transfer-encoding field is `chunked`,
https://golang.org/src/net/http/transfer.go#L178

- When a golang server reads the next request, the standard library parses the transfer-encoding field, passes the result to `Request.TransferEncoding`, and deletes the transfer-encoding field in the header, 
https://golang.org/src/net/http/transfer.go#L433,
The standard library actually considers it’s an error if the transfer-encoding field contains value other than `identity` or `chunked`,
https://golang.org/src/net/http/transfer.go#L447

Comment (caesarxuchao) [inline pkg/apiserver/proxy_test.go]: Asked the question in golang-nuts group:
https://groups.google.com/forum/#!topic/golang-nuts/g7xJk09xBxU

File: pkg/apiserver/proxy_test.go [hunk 8]
```diff
@@ -34,6 +36,124 @@ import (
...
+				contentLengthInHeader, err := strconv.Atoi(req.Header.Get("Content-Length"))
+				if err != nil {
+					t.Errorf("unexpected error: %v", err)
+				}
+				if e, a := expectedContentLength, int64(contentLengthInHeader); e != a {
+					t.Errorf("expected %v, got %v", e, a)
+				}
+			}
+
+			// The http library will strip the "Transfer-Encoding" field from request.Header.
```
Comment (caesarxuchao) [inline pkg/apiserver/proxy_test.go]: As the comment reads, Go library strips the "Transfer-Encoding" field. It will stores the value in request.TransferEncodings after some processing, the only possible value will be either "" or "chunked".

People say usually "gzip" will be a "Content-Encoding", so hopefully we won't get bit for this.

The discussion in go-nuts group: https://groups.google.com/forum/#!topic/golang-nuts/g7xJk09xBxU

File: pkg/apiserver/proxy_test.go [hunk 9]
```diff
@@ -34,6 +36,124 @@ import (
...
+			if e, a := "", req.Header.Get("Transfer-Encoding"); e != a {
+				t.Errorf("expected %v, got %v", e, a)
+			}
+			if e, a := item.contentEncoding, req.Header.Get("Content-Encoding"); e != a {
```
Comment (caesarxuchao) [inline pkg/apiserver/proxy_test.go]: The Go library doesn't change the "Content-Encoding" field.

File: pkg/apiserver/proxy_test.go (hunk 8 above)
Comment (lavalamp) [inline pkg/apiserver/proxy_test.go]: Thanks for the investigation!

Comment (liggitt) [inline pkg/apiserver/proxy_test.go]: yes, thanks... we should test with `content-encoding: gzip, transfer-encoding:''` as well

File: pkg/apiserver/proxy_test.go [hunk 10]
```diff
@@ -34,6 +36,125 @@ import (
...
+	}{
+		{[]string{}, "", "", "default"},
+		{[]string{"identity"}, "", "question", "default"},
+		{[]string{"chunked"}, "", "question", "default"},
+		// RFC2616 section-3.6: Whenever a transfer-coding is applied to a message-body, the set of
+		// transfer-codings MUST include "chunked", unless the message is terminated by closing the
+		// connection.
+		{[]string{"chunked", "gzip"}, "gzip", "qqqqqqqqqquestion", "default"},
+		{[]string{"chunked"}, "gzip", "qqqqqqqqqquestion", "default"},
+		{[]string{}, "gzip", "qqqqqqqqqquestion", "default"},
```
Comment (caesarxuchao) [inline pkg/apiserver/proxy_test.go]: @liggitt added two more test cases. PTAL.

**Comment by liggitt** [comment]
After digging into the http DefaultClient/DefaultTransport more, it looks like there is no way to make it faithfully echo transfer-encoding headers to the server. For low-level tests like this, I think we need to write headers and request bodies ourselves. Opened https://github.com/caesarxuchao/kubernetes/pull/2 to do so

File: pkg/apiserver/proxy.go (hunk 1 above)
Comment (liggitt) [inline pkg/apiserver/proxy.go]: I'm pretty sure we should only be setting ContentLength. TransferEncoding is interpreted, and the request body wrapped in an interpreting reader before we get it (that's why a request sent as "chunked" comes in with no transfer encoding, and reading the body returns unchunked data).

**Comment by caesarxuchao** [comment]
> it looks like there is no way to make it faithfully echo transfer-encoding/content-length headers to the server

That's true. I'll take a look at the PR. Thanks.

File: pkg/apiserver/proxy.go (hunk 1 above)
Comment (caesarxuchao) [inline pkg/apiserver/proxy.go]: But we need to keep the new request being "chunked" when it hits wire, right? If so, I think we need to copy the Request.TransferEncoding

The transfer writer will set the transfer-encoding header field to "chunked" if Request.TransferEncoding contains "chunked:
https://golang.org/src/net/http/transfer.go#L177
Same for the Body, the body will be chunked if Request.TransferEncoding contains "chunked":
https://golang.org/src/net/http/transfer.go#L213

Comment (liggitt) [inline pkg/apiserver/proxy.go]: hmm, I suppose so... I'm pretty sure the http package will always unwrap and transfer-encodings that are present, but in case they don't, it doesn't hurt to copy what remains to the new request

Comment (caesarxuchao) [inline pkg/apiserver/proxy.go]: tl;dr: we can keep this line, though it may not be useful.

After more digging, I found actually the http package is powerful enough deduce the outbound request should be "chunked" based on the fact that the content-length is left unset && there is content:
https://golang.org/src/net/http/transfer.go#L88

Well, it doesn't hurt to keep the copy in the PR, and *perhaps* it will save a broken request where "transfer-encoding:chunked" and "content-length: non-zero-value" are set at the same time.

File: pkg/apiserver/proxy_test.go [hunk 11]
```diff
@@ -34,6 +38,223 @@ import (
...
+		defer conn.Close()
+
+		// Add standard http 1.1 headers
+		if item.reqHeaders == nil {
+			item.reqHeaders = http.Header{}
+		}
+		item.reqHeaders.Add("Connection", "close")
+		item.reqHeaders.Add("Host", server.Listener.Addr().String())
+
+		// We directly write to the connection to bypass the Go library's manipulation of the Request.Header.
```
Comment (caesarxuchao) [inline pkg/apiserver/proxy_test.go]: @liggitt I added a comment here. Do you think it's appropriate?

File: pkg/registry/generic/rest/proxy.go [hunk 12]
```diff
@@ -114,6 +114,10 @@ func (h *UpgradeAwareProxyHandler) ServeHTTP(w http.ResponseWriter, req *http.Re
 		return
 	}
 	newReq.Header = req.Header
+	newReq.ContentLength = req.ContentLength
+	// Copy the TransferEncoding is for future-proofing. Currently Go only supports "chunked" and
+	// it can determine the TransferEncoding based on ContentLength and the Body.
+	newReq.TransferEncoding = req.TransferEncoding
```
Comment (caesarxuchao) [inline pkg/registry/generic/rest/proxy.go]: @liggitt I added back this for future-proofing.

File: pkg/apiserver/proxy_test.go (hunk 11 above)
Comment (liggitt) [inline pkg/apiserver/proxy_test.go]: sure

**Comment by liggitt** [comment]
LGTM, feel free to squash my commit in

**Comment by liggitt** [comment]
> Maybe different version of go prints a slightly different error message? Maybe you should look into parsing the response with the tools in net/http instead of string comparison?

We're not checking error messages, we're making sure the response ended with what the backend server output. That should always be an identical match

**Comment by ixdy** [comment]
There was some Jenkins weirdness earlier. Let's retest to make sure that was not the cause:
@k8s-bot test this please

**Comment by caesarxuchao** [comment]
I think we just need a rebase...I forgot deads2k's PR which I reviewed changed the url.

**Comment by caesarxuchao** [comment]
Squashed. I kept liggitt in the commit message as an acknowledgement to his contribution.

**Comment by caesarxuchao** [comment]
@lavalamp PTAL. Thanks.

**Comment by caesarxuchao** [comment]
@k8s-bot test this please

**Comment by lavalamp** [comment]
LGTM, but the code duplication is unfortunate. I guess we already had that problem. :)

**Comment by wojtek-t** [comment]
@k8s-bot unit test this please

**Comment by caesarxuchao** [comment]
@k8s-bot unit test this please

**Comment by caesarxuchao** [comment]
Unrelated test failure:
```
+++ [1203 19:55:59] Testing kubectl(:multiple resources)
Testing with file hack/testdata/multi-resource-yaml.yaml and replace with file hack/testdata/multi-resource-yaml-modify.yaml
Successful get services {{range.items}}{{.metadata.name}}:{{end}}: kubernetes:

FAIL!
Get rc {{range.items}}{{.metadata.name}}:{{end}}
  Expected: 
  Got:      nginx-deployment-p889s:

1058 ./hack/test-cmd.sh

!!! Error in ./hack/test-cmd.sh:50
```

**Comment by caesarxuchao** [comment]
@k8s-bot test this please

**Comment by caesarxuchao** [comment]
unrelated e2e test failure:
```
Kubernetes e2e suite.Kubectl client Kubectl describe should check if kubectl describe prints relevant information for rc and pods 
```

**Comment by caesarxuchao** [comment]
@k8s-bot e2e test this please
